### ISSUE TITLE
Implement Python bindings

### ISSUE THREAD

**Issue by andygrove** [ballista-compute/ballista#418]
DataFusion has Python bindings available [1] so it should be possible to adapt this code and add Python bindings for Ballista.

[1] https://pypi.org/project/datafusion/

**Comment by djKooks** [comment]
@andygrove interested in here also. How do you think to start in?

**Comment by andygrove** [comment]
@jorgecarleitao if you have time, would you be able to offer guidance on where to get started with this task?

**Comment by jorgecarleitao** [comment]
Hey @djKooks ! Great to hear!

So, if I understand correctly, we would like something like

```python
ctx = ballista.local_context()

df = ctx.read_parquet(path: str)

df.select(...).agg(['a'], count('b'))

batches = df.collect()
```

as an initial goal.

My experience with `python-datafusion` and `pyo3` is that we will need to create structs that essentially bridge the Rust client and Python. For example, for datafusion's `ExecutionContext`, I did it like [this](https://github.com/jorgecarleitao/datafusion-python/blob/master/src/context.rs#L22).

The second step is to offer methods that map a python method to the corresponding rust method. For example

[here](https://github.com/jorgecarleitao/datafusion-python/blob/master/src/context.rs#L47):

```rust
fn create_dataframe(
        &mut self,
        partitions: Vec<Vec<PyObject>>,
        py: Python,
    ) -> PyResult<dataframe::DataFrame> {...}
```

The important part to notice here is that both the arguments and the output are python-like objects (that `pyo3` maps to rust and vice-versa).

For the DataFrame API itself, I had to create a small DataFrame API in Python that maps to the rust DataFrame API, [here](https://github.com/jorgecarleitao/datafusion-python/blob/master/src/expression.rs#L10).

My suggestion is to fork python-datafusion, that already has CI, packaging, tests, and all that stuff, modify `Cargo.toml` to depend on `Ballista`, and replace the execution Context from DataFusion by the Ballista one, fixing compilation errors until it runs. It may just work ^_^

If this works, I can try to split python-datafusion in two parts or try to publish it as a Rust lib that can be consumed by ballista. One idea is to introduce a `ExecutionContext` trait that can be plugged in on generics, so that we do not need to copy-paste code around.
